api/lib/opentelemetry/context.rb

```diff
@@ -29,23 +29,44 @@ def current
         Thread.current[KEY] ||= ROOT
       end
 
-      # Sets the current context
-      #
-      # @param [Context] ctx The context to be made active
+      # @api private
       def current=(ctx)
         Thread.current[KEY] = ctx
       end
 
+      # Returns the previous context so that it can be restored
+      #
+      # @param [Context] context The new context
+      # @return [Context] prev The previous context
+      def attach(context)
+        return current.parent if context == current
```

Comment:
This is an interesting corner case 🤔 - it is weirdly open to abuse, in that you can use it to retrieve the parent of the current context. Should we document this?

Actually, this makes me mildly concerned about our current implementation with the stack of contexts represented as an internally linked list. With other implementations (e.g. I think Java maintains the context stack as an array), it is possible for a context to be pushed (attached) twice. Our implementation won't allow that, which could result in some legitimate code patterns 💥 .

Reply:
This corner case was to prevent repeatedly attaching the same context and (due to the way I implemented this) causing a circular dependency that would prevent users from properly detaching the context.

I agree that this is concerning, I was admittedly a little too focused on keeping parity with the previous implementation but with making attach/detach public it certainly changes considerations.

I suppose we want to support attaching the same context multiple times?  Otherwise it's too easy to have your attach/detach calls fall out of sync.

Reply:
> I suppose we want to support attaching the same context multiple times? Otherwise it's too easy to have your attach/detach calls fall out of sync.

Unless the spec forbids it, I think that'd be a nice thing to support.